### PR TITLE
Add REST API to add, remove and update CustomizedStateAggregationConfig

### LINKED ISSUE #791: TODO: Add relevant REST Level APIs for add/remove CustomizedViewConfig
The REST APIs for adding and removing CustomizedViewConfig need to be implemented.

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Fixes #791 

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
In this commit the below REST APIs have been added.
1- addCustomizedStateAggregationConfig
2- removeCustomizedStateAggregationConfig
3- getCustomizedStateAggregationConfig
4- updateCustomizedStateAggregationConfig
Tests have been added to check the functionality of these REST APIs.
Also some of the deprecated calls have been updated.

- [x] The following tests are written for this issue:
TestClusterAccessor.testAddCustomizedConfigNonExistedCluster
TestClusterAccessor.testAddCustomizedConfig
TestClusterAccessor.testDeleteCustomizedConfig
TestClusterAccessor.testUpdateCustomizedConfig

- [x] The following is the result of the "mvn test" command on the appropriate module:

Helix-core:
[INFO] Tests run: 908, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3,535.183 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 908, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  59:00 min
[INFO] Finished at: 2020-02-21T16:16:31-08:00
[INFO] ------------------------------------------------------------------------

Helix-rest:
[INFO] Tests run: 97, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 25.632 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 97, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  30.507 s
[INFO] Finished at: 2020-02-21T15:17:01-08:00
[INFO] ------------------------------------------------------------------------

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality
- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 1]
```diff
@@ -267,6 +268,129 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
     return JSONRepresentation(config.getRecord());
   }
 
+
+  @PUT
+  @Path("{clusterId}/customizedstateaggregationconfig")
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: This endpoint name is so difficult to read/write. Can you change it to snake-case: `customized-state-xxx`?

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 2]
```diff
@@ -267,6 +268,129 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+    }
+
+    try {
+      CustomizedStateAggregationConfig customizedStateAggregationConfig =
+          new CustomizedStateAggregationConfig.Builder(record).build();
+      admin.addCustomizedStateAggregationConfig(clusterId, customizedStateAggregationConfig);
+    } catch (HelixException ex) {
+      _logger.error("Error in adding a CustomizedStateAggregationConfig to cluster: " + clusterId,
+          ex);
+      return badRequest(ex.getMessage());
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: For this exception, is it caused by a server error or a client error? It is good to make sure the response status code is appropriate.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 3]
```diff
@@ -267,6 +268,129 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+      } catch (HelixException ex) {
+        return badRequest(ex.getMessage());
+      }
+    }
+
+    HelixAdmin admin = getHelixAdmin();
+
+    try {
+      switch (command) {
+      case delete: {
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: I would remove `{}` for `case` blocks.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 4]
```diff
@@ -267,6 +268,129 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+    HelixZkClient zkClient = getHelixZkClient();
+    if (!ZKUtil.isClusterSetup(clusterId, zkClient)) {
+      return notFound();
+    }
+
+    // Here to update cloud config
+    Command command;
+    if (commandStr == null || commandStr.isEmpty()) {
+      command = Command.update; // Default behavior
+    } else {
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: No need for else.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 5]
```diff
@@ -267,6 +268,129 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+
+    return OK();
+  }
+
+  @GET
+  @Path("{clusterId}/customizedstateaggregationconfig")
+  public Response getCustomizedStateAggregationConfig(@PathParam("clusterId") String clusterId) {
+    HelixZkClient zkClient = getHelixZkClient();
+    if (!ZKUtil.isClusterSetup(clusterId, zkClient)) {
+      return notFound();
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: I would suggest adding an error message to the response.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 6]
```diff
@@ -267,6 +268,129 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+    return OK();
+  }
+
+  @DELETE
+  @Path("{clusterId}/customizedstateaggregationconfig")
+  public Response removeCustomizedStateAggregationConfig(@PathParam("clusterId") String clusterId) {
+    HelixZkClient zkClient = getHelixZkClient();
+    if (!ZKUtil.isClusterSetup(clusterId, zkClient)) {
+      return notFound("Cluster is not properly setup!");
+    }
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: These lines could be in a private check method to reduce duplicate code.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 7]
```diff
@@ -267,6 +268,129 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+        configAccessor.getCustomizedStateAggregationConfig(clusterId);
+
+    if (customizedStateAggregationConfig != null) {
+      return JSONRepresentation(customizedStateAggregationConfig.getRecord());
+    }
+
+    return notFound();
+  }
+
+  @POST
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: I just followed the convention here. PUT has been used to add a completely new config. DELETE for complete deletion of the config. POST for update the information.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 6 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Ok. Fixed.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 5 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Done. Thanks.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 4 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: It is needed. If commandStr == null or commandStr.isEmpty() then we want to continued with the request and use update as default. We do not want to return badRequest.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 3 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Done.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 2 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Actually this exception is not needed. Removed it.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 1 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Fixed.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 8]
```diff
@@ -267,6 +268,115 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+    HelixZkClient zkClient = getHelixZkClient();
+    if (!doesClusterExist(clusterId)) {
+      return notFound(String.format("Cluster %s does not exist", clusterId));
+    }
+
+    HelixAdmin admin = getHelixAdmin();
+    try {
+      admin.removeCustomizedStateAggregationConfig(clusterId);
+    } catch (Exception ex) {
+      _logger.error("Cannot remove CustomizedStateAggregationConfig to cluster: " + clusterId, ex);
```
Comment (zhangmeng916) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: NIP: to -> from

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 9]
```diff
@@ -267,6 +268,115 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+    }
+
+    HelixAdmin admin = getHelixAdmin();
+
+    try {
+      switch (command) {
+      case delete:
+        admin.removeTypeFromCustomizedStateAggregationConfig(clusterId, type);
+        break;
+      case update:
```
Comment (zhangmeng916) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: The update here is confusing, especially this command is also called "update". Actually this is "add", right?

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 8 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Fixed.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 9 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: I added a new type with the name Add.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 10]
```diff
@@ -267,6 +268,115 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+      return serverError(ex);
+    }
+
+    return OK();
+  }
+
+  @DELETE
+  @Path("{clusterId}/customized-state-aggregation-config")
+  public Response removeCustomizedStateAggregationConfig(@PathParam("clusterId") String clusterId) {
+    HelixZkClient zkClient = getHelixZkClient();
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Unused.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 11]
```diff
@@ -267,6 +268,115 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+    } catch (IOException e) {
+      _logger.error("Failed to deserialize user's input " + content + ", Exception: " + e);
+      return badRequest("Input is not a vaild ZNRecord!");
+    }
+
+    try {
+      CustomizedStateAggregationConfig customizedStateAggregationConfig =
+          new CustomizedStateAggregationConfig.Builder(record).build();
+      admin.addCustomizedStateAggregationConfig(clusterId, customizedStateAggregationConfig);
+    } catch (Exception ex) {
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Can we not catch a more specific exception? It seems there could be runtime HelixException?

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 12]
```diff
@@ -267,6 +268,115 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+    if (!doesClusterExist(clusterId)) {
+      return notFound(String.format("Cluster %s does not exist", clusterId));
+    }
+
+    HelixAdmin admin = getHelixAdmin();
+    ZNRecord record;
+    try {
+      record = toZNRecord(content);
+    } catch (IOException e) {
+      _logger.error("Failed to deserialize user's input " + content + ", Exception: " + e);
```
Comment (pkuwm) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: This is a user input/client error, not a server error.  We don't have to log this especially on error level in server.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/AbstractResource.java [hunk 13]
```diff
@@ -62,6 +62,7 @@
     enablePartitions,
     disablePartitions,
     update,
+    Add,
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/AbstractResource.java]: Fixed.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 10 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Removed.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 12 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Removed the log.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 11 above)
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Thanks for the comment. I still prefer to catch all the exception like we do in other functions.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 14]
```diff
@@ -362,7 +362,7 @@ public Response updateCustomizedStateAggregationConfig(@PathParam("clusterId") S
       case delete:
         admin.removeTypeFromCustomizedStateAggregationConfig(clusterId, type);
         break;
-      case update:
+      case Add:
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Fixed.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 15]
```diff
@@ -267,6 +268,115 @@ public Response getClusterConfig(@PathParam("clusterId") String clusterId) {
...
+        admin.removeTypeFromCustomizedStateAggregationConfig(clusterId, type);
+        break;
+      case Add:
+        admin.addTypeToCustomizedStateAggregationConfig(clusterId, type);
+        break;
+      default:
+        return badRequest("Unsupported command " + commandStr);
+      }
+    } catch (Exception ex) {
+      _logger.error("Failed to " + command + " CustomizedStateAggregationConfig for cluster " + clusterId + " new type: "
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Done.

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @zhangmeng916.

Add REST API to add, remove and update CustomizedStateAggregationConfig

In this commit the below REST APIs have been added.
1- addCustomizedStateAggregationConfig
2- removeCustomizedStateAggregationConfig
3- getCustomizedStateAggregationConfig
4- updateCustomizedStateAggregationConfig
Tests have been added to check the functionality of these REST APIs.
Also some of the deprecated calls have been updated.

**Pull request merged**